### ISSUE TITLE
geo/geomfn: implement ST_ClusterWithin({geometry,float8})

### ISSUE THREAD

**Issue by otan** [cockroachdb/cockroach#48901]
Implement `ST_ClusterWithin` on arguments {geometry,float8}, which should follow the [PostGIS implementation](https://postgis.net/docs/ST_ClusterWithin.html).

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geogfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geomfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output manually. Tests here should just ensure the builtin is linked end to end.
* Ensure the documentation is regenerated by calling `make buildshort`.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48552 for an example PR.

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2020-05-14T15:32:08Z. Changes to titles, body and labels may be overwritten.</sub>

**Comment by klesniewski** [comment]
We are migrating parts of our application to CockroachDB and we need ST_ClusterWithin to implement displaying assets on the map.

![image](https://user-images.githubusercontent.com/6713429/98788905-cd212200-2401-11eb-8bf9-e8ad31108bdc.png)

Typically tenants within our system have hundreds to tens of thousands of assets. Unfortunately, putting all tenant's assets on the map individually doesn't work well - server response would be big (or spread over many pages) and client side would have a lot of work to render the map what results in poor user experience. Therefore, our backend takes care of clustering and returns clusters of assets. ST_ClusterWithin would help us by pushing the work of computing clusters to the database engine.

**Comment by otan** [comment]
Thanks @klesniewski - are there any other clustering algorithms that you find you commonly need?

**Comment by klesniewski** [comment]
At the moment this is the only use case we have for clustering in transactional processing. I don't have much experience with clustering algorithms in Postgres, but from what I read ST_ClusterWithin sounded closest to what we need. The other ones considered were [ST_ClusterDBSCAN](https://postgis.net/docs/ST_ClusterDBSCAN.html) and [ST_ClusterKMeans](https://postgis.net/docs/ST_ClusterKMeans.html) (as compared [here](https://gis.stackexchange.com/a/11778/131455)).

Basically it is not as important for us how exactly the points are clustered together as long it looks quite intuitively in UI. What matters is that ideally we can return for each cluster a number of records in it, and if the cluster is of small size (let's say 1 or 3), then actually have the record data, so that we can put it on the map unclustered.

**Comment by klesniewski** [comment]
A little update on our use case.

We have looked into different methods and ruled out DBSCAN and KMeans. We have ruled out KMeans as it requires to specify number of clusters we want to find. We have ruled out DBSCAN because it is not kind of clustering we are looking here for. DBSCAN can cluster non-linearly separable clusters, which is a nice property, but nor desirable in this use case:
![image](https://user-images.githubusercontent.com/6713429/100267588-f9858400-2f53-11eb-84be-520bc499ab30.png)

According to article [DBSCAN Clustering in PostGIS](http://www.danbaston.com/posts/2016/06/02/dbscan-clustering-in-postgis.html), ST_ClusterWithin is a special case of DBSCAN, so as such we have ruled it out too.

We are currently implementing a solution based on geohash, roughly as described [here](https://tech.willhaben.at/geo-clustering-3-000-000-points-on-the-fly-a-brief-how-to-9f04d8d5b3a7), [here](https://blog.trifork.com/2013/08/01/server-side-clustering-of-geo-points-on-a-map-using-elasticsearch/) or [here](https://www.slideshare.net/EvgeniyKhist/serverside-geoclustering-based-on-geohash).
